// extra-logger/lib/transports/terminal-transport.js
import { Level } from "../types.js";
import chalk from 'chalk';
import { isUndefined, isntUndefined } from '@blackglory/prelude';
export class TerminalTransport {
    constructor(options = {}) {
        this.options = options;
        this.id = 0;
    }
    send(message) {
        if (isUndefined(message.elapsedTime)) {
            console.log(this.stringifyMessage(message));
        }
        else {
            if (isUndefined(this.options.logMinimumDuration)) {
                console.log(this.stringifyMessage(message));
            }
            else {
                if (message.elapsedTime >= this.options.logMinimumDuration) {
                    console.log(this.stringifyMessage(message));
                }
            }
        }
    }
    stringifyMessage(message) {
        const str = [
            `[${levelToString(message.level)}]`
                + `[${formatDate(message.timestamp)}]`,
            `#${this.createId()}`
        ];
        if (isntUndefined(message.namespace)) {
            str.push(`${formatNamespace(message.namespace)}`);
        }
        str.push(`${message.message}`);
        if (isntUndefined(message.elapsedTime)) {
            str.push(`${formatElapsedTime(message.elapsedTime)}`);
        }
        return str.join(' ');
    }
    createId() {
        return ++this.id;
    }
}
function levelToString(level) {
    switch (level) {
        case Level.Info: return 'INFO';
        case Level.Debug: return 'DEBUG';
        case Level.Warn: return 'WARN';
        case Level.Trace: return 'TRACE';
        case Level.Error: return 'ERROR';
        case Level.Fatal: return 'FATAL';
        default: return 'NONE';
    }
}
function formatNamespace(namespace) {
    return chalk.cyan(namespace);
}
function formatDate(timestamp) {
    return new Date(timestamp).toISOString();
}
function formatElapsedTime(elapsed) {
    if (elapsed <= 100)
        return chalk.green(`${elapsed}ms`);
    if (elapsed <= 300)
        return chalk.yellow(`${elapsed}ms`);
    return chalk.red(`${elapsed}ms`);
}
