// @blackglory/structures/lib/emitter.js
import { go } from '@blackglory/go';
export class Emitter {
    constructor() {
        this.map = new Map();
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    on(event, listener) {
        const set = go(() => {
            const set = this.map.get(event);
            if (set) {
                return set;
            }
            else {
                const set = new Set();
                this.map.set(event, set);
                return set;
            }
        });
        set.add(handler);
        return () => {
            const handlers = this.map.get(event);
            if (handlers) {
                const deleted = handlers.delete(handler);
                if (deleted && handlers.size === 0) {
                    this.map.delete(event);
                }
            }
        };
        function handler(...args) {
            listener(...args);
        }
    }
    once(event, listener) {
        const removeListener = this.on(event, (...args) => {
            listener(...args);
            removeListener();
        });
        return removeListener;
    }
    emit(event, ...args) {
        var _a;
        (_a = this.map.get(event)) === null || _a === void 0 ? void 0 : _a.forEach(cb => cb(...args));
    }
    removeAllListeners(event) {
        var _a;
        (_a = this.map.get(event)) === null || _a === void 0 ? void 0 : _a.clear();
    }
}
