// extra-request/lib/request.js
import { Headers, Request } from 'extra-fetch';
import { pipeRequestOptionsTransformers } from "./pipe-request-options-transformers.js";
export function get(...transformers) {
    return request('GET', ...transformers);
}
export function head(...transformers) {
    return request('HEAD', ...transformers);
}
export function post(...transformers) {
    return request('POST', ...transformers);
}
export function put(...transformers) {
    return request('PUT', ...transformers);
}
export function patch(...transformers) {
    return request('PATCH', ...transformers);
}
export function del(...transformers) {
    return request('DELETE', ...transformers);
}
export function request(method, ...transformers) {
    const requestOptions = pipeRequestOptionsTransformers(...transformers);
    const headers = new Headers(requestOptions.headers);
    return new Request(requestOptions.url.href, {
        method,
        headers,
        signal: requestOptions.signal,
        body: requestOptions.payload,
        keepalive: requestOptions.keepalive,
        redirect: requestOptions.redirect
    });
}
