// iterable-operator/lib/reduce-async.js
import { isUndefined } from 'extra-utils';
import { isAsyncIterable } from "./is-async-iterable.js";
export function reduceAsync(iterable, fn, initialValue) {
    if (isUndefined(initialValue)) {
        return reduceAsyncWithoutInitialValue(iterable, fn);
    }
    else {
        return reduceAsyncWithInitialValue(iterable, fn, initialValue);
    }
}
async function reduceAsyncWithInitialValue(iterable, fn, initialValue) {
    let result = initialValue, index = 0;
    for await (const currentValue of iterable) {
        result = await fn(result, currentValue, index++);
    }
    return result;
}
function reduceAsyncWithoutInitialValue(iterable, fn) {
    if (isAsyncIterable(iterable)) {
        return reduceAsyncIterable(iterable);
    }
    else {
        return reduceIterable(iterable);
    }
    async function reduceAsyncIterable(iterable) {
        var _a;
        const iterator = iterable[Symbol.asyncIterator]();
        let done;
        try {
            let result = await readInitialValue(iterator);
            let index = 1;
            let value;
            while ({ value, done } = await iterator.next(), !done) {
                result = await fn(result, value, index++);
            }
            return result;
        }
        finally {
            if (!done)
                await ((_a = iterator.return) === null || _a === void 0 ? void 0 : _a.call(iterator));
        }
        async function readInitialValue(iterator) {
            const result = await iterator.next();
            if (result.done) {
                done = true;
                throw new Error('Reduce of empty iterable with no initial value');
            }
            return result.value;
        }
    }
    async function reduceIterable(iterable) {
        var _a;
        const iterator = iterable[Symbol.iterator]();
        let done;
        try {
            let result = readInitialValue(iterator);
            let index = 1;
            let value;
            while ({ value, done } = iterator.next(), !done) {
                result = await fn(result, value, index++);
            }
            return result;
        }
        finally {
            if (!done)
                (_a = iterator.return) === null || _a === void 0 ? void 0 : _a.call(iterator);
        }
        function readInitialValue(iterator) {
            const result = iterator.next();
            if (result.done) {
                done = true;
                throw new Error('Reduce of empty iterable with no initial value');
            }
            return result.value;
        }
    }
}
