// @blackglory/structures/lib/sparse-set.js
export class SparseSet {
    constructor(array = []) {
        this.indexToValue = [];
        const valueToIndex = [];
        if (array.length > 0) {
            for (const [index, value] of array.entries()) {
                valueToIndex[value] = index;
            }
        }
        this.valueToIndex = valueToIndex;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    [Symbol.iterator]() {
        return this.indexToValue[Symbol.iterator]();
    }
    get size() {
        return this.indexToValue.length;
    }
    values() {
        return this.indexToValue[Symbol.iterator]();
    }
    has(value) {
        return value in this.valueToIndex;
    }
    add(value) {
        if (!this.has(value)) {
            const index = this.indexToValue.length;
            this.indexToValue.push(value);
            this.valueToIndex[value] = index;
        }
    }
    delete(value) {
        if (this.has(value)) {
            const lastValue = this.indexToValue.pop();
            if (value === lastValue) {
                delete this.valueToIndex[value];
            }
            else {
                const index = this.valueToIndex[value];
                this.indexToValue[index] = lastValue;
                this.valueToIndex[lastValue] = index;
                delete this.valueToIndex[value];
            }
            return true;
        }
        else {
            return false;
        }
    }
    clear() {
        this.indexToValue.length = 0;
        this.valueToIndex.length = 0;
    }
}
