// value-getter/lib/value-getter.js
import { isFunction, isNull, isUndefined } from 'extra-utils';
export class ValueGetter {
    constructor(...args) {
        if (args.length === 1) {
            const [getter] = args;
            this.getter = getter;
            this.name = 'anonymous';
        }
        else {
            const [name, getter] = args;
            this.getter = getter;
            this.name = name;
        }
    }
    default(val) {
        return new ValueGetter(this.name, () => { var _a; return (_a = this.value()) !== null && _a !== void 0 ? _a : val; });
    }
    assert(assert) {
        return new ValueGetter(this.name, () => {
            const val = this.value();
            assert(val);
            return val;
        });
    }
    required() {
        return this.assert(val => {
            if (isUndefined(val) || isNull(val)) {
                throw new Error(`${this.name} should not be null or undefined`);
            }
        });
    }
    memoize(param) {
        const get = () => this.value();
        return new ValueGetter(this.name, () => {
            const cache = isFunction(param) ? param() : param;
            if (cache.has(get)) {
                return cache.get(get);
            }
            else {
                const val = get();
                cache.set(get, val);
                return val;
            }
        });
    }
    convert(convert) {
        return new ValueGetter(this.name, () => convert(this.getter()));
    }
    tap(sideEffect) {
        return new ValueGetter(this.name, () => {
            const val = this.value();
            sideEffect(val);
            return val;
        });
    }
    get() {
        return this.getter;
    }
    value() {
        return this.getter();
    }
}
