// @blackglory/structures/lib/trie-map.js
import { zip, toArray, map } from 'iterable-operator';
import { isntUndefined } from 'extra-utils';
class TrieNode {
    constructor(value) {
        this.value = value;
        this.children = new Map();
    }
}
export class TrieMap {
    constructor() {
        this.root = new TrieNode();
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    *entries() {
        yield* dfs(this.root, []);
        function* dfs(node, path) {
            for (const [subPath, childNode] of node.children) {
                const newPath = [...path, subPath];
                if (isntUndefined(childNode.value)) {
                    yield [newPath, childNode.value];
                }
                yield* dfs(childNode, newPath);
            }
        }
    }
    keys() {
        return map(this.entries(), ([key]) => key);
    }
    values() {
        return map(this.entries(), ([_, value]) => value);
    }
    set(key, value) {
        let node = this.root;
        for (const unitOfKey of key) {
            if (!node.children.has(unitOfKey)) {
                node.children.set(unitOfKey, new TrieNode());
            }
            node = node.children.get(unitOfKey);
        }
        node.value = value;
        return this;
    }
    has(key) {
        let node = this.root;
        for (const unitOfKey of key) {
            if (node.children.has(unitOfKey)) {
                node = node.children.get(unitOfKey);
            }
            else {
                return false;
            }
        }
        return node.value !== undefined;
    }
    get(key) {
        let node = this.root;
        for (const unitOfKey of key) {
            if (node.children.has(unitOfKey)) {
                node = node.children.get(unitOfKey);
            }
            else {
                return undefined;
            }
        }
        return node.value;
    }
    delete(key) {
        const parentNodes = [];
        let node = this.root;
        for (const unitOfKey of key) {
            if (node.children.has(unitOfKey)) {
                parentNodes.push(node);
                node = node.children.get(unitOfKey);
            }
            else {
                return false;
            }
        }
        delete node.value;
        if (node.children.size === 0) {
            for (const [unitOfKey, parentNode] of toArray(zip(key, parentNodes)).reverse()) {
                parentNode.children.delete(unitOfKey);
                if (parentNode.children.size !== 0)
                    break;
            }
        }
        return true;
    }
}
