// @blackglory/structures/lib/string-trie-map.js
import { zip, toArray, map } from 'iterable-operator';
import { isntUndefined } from 'extra-utils';
class TrieNode {
    constructor(value) {
        this.value = value;
        this.children = new Map();
    }
}
export class StringTrieMap {
    constructor() {
        this.root = new TrieNode();
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    entries() {
        return dfs(this.root, '');
        function* dfs(node, path) {
            for (const [char, childNode] of node.children) {
                const newPath = path + char;
                if (isntUndefined(childNode.value)) {
                    yield [newPath, childNode.value];
                }
                yield* dfs(childNode, newPath);
            }
        }
    }
    keys() {
        return map(this.entries(), ([key]) => key);
    }
    values() {
        return map(this.entries(), ([_, value]) => value);
    }
    set(key, value) {
        let node = this.root;
        for (const char of key) {
            if (!node.children.has(char)) {
                node.children.set(char, new TrieNode());
            }
            node = node.children.get(char);
        }
        node.value = value;
        return this;
    }
    has(key) {
        let node = this.root;
        for (const char of key) {
            if (node.children.has(char)) {
                node = node.children.get(char);
            }
            else {
                return false;
            }
        }
        return node.value !== undefined;
    }
    get(key) {
        let node = this.root;
        for (const char of key) {
            if (node.children.has(char)) {
                node = node.children.get(char);
            }
            else {
                return undefined;
            }
        }
        return node.value;
    }
    delete(key) {
        const parentNodes = [];
        let node = this.root;
        for (const char of key) {
            if (node.children.has(char)) {
                parentNodes.push(node);
                node = node.children.get(char);
            }
            else {
                return false;
            }
        }
        delete node.value;
        if (node.children.size === 0) {
            for (const [char, parentNode] of toArray(zip(key, parentNodes)).reverse()) {
                parentNode.children.delete(char);
                if (parentNode.children.size !== 0)
                    break;
            }
        }
        return true;
    }
}
