// @blackglory/structures/lib/typed-sparse-set.js
export class TypedSparseSet {
    valueToIndex;
    indexToValue;
    constructor(array) {
        const valueToIndex = [];
        if (array.length > 0) {
            for (const [index, value] of array.internalTypedArray.entries()) {
                valueToIndex[value] = index;
            }
        }
        this.indexToValue = array;
        this.valueToIndex = valueToIndex;
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    [Symbol.iterator]() {
        return this.values();
    }
    get size() {
        return this.indexToValue.length;
    }
    *values() {
        for (let i = 0; i < this.indexToValue.length; i++) {
            yield this.indexToValue.internalTypedArray[i];
        }
    }
    has(value) {
        return this.valueToIndex[value] !== undefined;
    }
    add(value) {
        if (!this.has(value)) {
            const index = this.indexToValue.length;
            this.indexToValue.push(value);
            this.valueToIndex[value] = index;
        }
    }
    delete(value) {
        if (this.has(value)) {
            const lastValue = this.indexToValue.pop();
            if (value === lastValue) {
                this.valueToIndex[value] = undefined;
            }
            else {
                const index = this.valueToIndex[value];
                this.indexToValue.internalTypedArray[index] = lastValue;
                this.valueToIndex[lastValue] = index;
                this.valueToIndex[value] = undefined;
            }
            return true;
        }
        else {
            return false;
        }
    }
    clear() {
        this.valueToIndex.length = 0;
        this.indexToValue.clear();
    }
}
