// extra-defer/lib/async-generator-executor.js
import { goAsyncGenerator } from '@blackglory/go';
export class AsyncGeneratorExecutor {
    constructor({ autoClear = true } = {}) {
        this.callbacks = [];
        this.autoClear = autoClear;
    }
    get size() {
        return this.callbacks.length;
    }
    defer(callback) {
        this.callbacks.push(callback);
    }
    remove(callback) {
        this.callbacks = this.callbacks.filter(x => x !== callback);
    }
    clear() {
        this.callbacks = [];
    }
    async *execute(...args) {
        const callbacks = this.callbacks;
        if (this.autoClear) {
            this.clear();
        }
        for (const callback of this.iterate(callbacks)) {
            yield* goAsyncGenerator(() => callback(...args));
        }
    }
    async *executeSettled(...args) {
        const callbacks = this.callbacks;
        if (this.autoClear) {
            this.clear();
        }
        for (const callback of this.iterate(callbacks)) {
            try {
                yield* goAsyncGenerator(() => callback(...args));
            }
            catch (_a) {
            }
        }
    }
}
