// extra-defer/lib/executor.js
import { each } from 'extra-promise';
export class Executor {
    constructor({ autoClear = true } = {}) {
        this.callbacks = [];
        this.autoClear = autoClear;
    }
    get size() {
        return this.callbacks.length;
    }
    defer(callback) {
        this.callbacks.push(callback);
    }
    remove(callback) {
        this.callbacks = this.callbacks.filter(x => x !== callback);
    }
    clear() {
        this.callbacks = [];
    }
    async execute(...args) {
        await this.all(1, ...args);
    }
    async executeSettled(...args) {
        await this.allSettled(1, ...args);
    }
    async all(concurrency = Infinity, ...args) {
        const callbacks = this.callbacks;
        if (this.autoClear) {
            this.clear();
        }
        await each(this.iterate(callbacks), callback => callback(...args), concurrency);
    }
    async allSettled(concurrency = Infinity, ...args) {
        const callbacks = this.callbacks;
        if (this.autoClear) {
            this.clear();
        }
        await each(this.iterate(callbacks), async (callback) => {
            try {
                await callback(...args);
            }
            catch (_a) {
            }
        }, concurrency);
    }
}
