// extra-promise/lib/classes/mutable-deferred.js
import { Deferred } from "./deferred.js";
import { pass } from '@blackglory/pass';
export class MutableDeferred {
    constructor() {
        this.deferred = this.createDeferred();
        this.isFirstRun = true;
    }
    get then() {
        return this.deferred.then.bind(this.deferred);
    }
    resolve(value) {
        if (this.isFirstRun) {
            this.isFirstRun = false;
            this.deferred.resolve(value);
        }
        else {
            // 用新的Deferred覆盖, 下一个MutableDeferred的调用者会获得新值.
            const deferred = this.createDeferred();
            deferred.resolve(value);
            this.deferred = deferred;
        }
    }
    reject(reason) {
        if (this.isFirstRun) {
            this.isFirstRun = false;
            this.deferred.reject(reason);
        }
        else {
            // 用新的Deferred覆盖, 下一个MutableDeferred的调用者会获得新值.
            const deferred = this.createDeferred();
            deferred.reject(reason);
            this.deferred = deferred;
        }
    }
    createDeferred() {
        const deferred = new Deferred();
        Promise.resolve(deferred).catch(pass);
        return deferred;
    }
}
