// @blackglory/structures/lib/typed-sparse-map.js
import { assert } from '@blackglory/errors';
export class TypedSparseMap {
    keyToIndex = [];
    indexToKey = [];
    indexToValue;
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    get size() {
        return this.indexToKey.length;
    }
    get internalTypedArray() {
        return this.indexToValue.internalTypedArray;
    }
    constructor(array) {
        assert(array.length === 0, 'array should be empty');
        this.indexToValue = array;
    }
    *entries() {
        for (let i = 0; i < this.indexToKey.length; i++) {
            yield [
                this.indexToKey[i],
                this.indexToValue.internalTypedArray[i]
            ];
        }
    }
    keys() {
        return this.indexToKey.values();
    }
    *values() {
        for (let i = 0; i < this.indexToKey.length; i++) {
            yield this.indexToValue.internalTypedArray[i];
        }
    }
    getInternalIndexOfKey(key) {
        return this.keyToIndex[key];
    }
    has(key) {
        return this.keyToIndex[key] !== undefined;
    }
    get(key) {
        if (this.has(key)) {
            const index = this.keyToIndex[key];
            return this.indexToValue.internalTypedArray[index];
        }
        else {
            return undefined;
        }
    }
    set(key, value) {
        if (this.has(key)) {
            const index = this.keyToIndex[key];
            this.indexToValue.internalTypedArray[index] = value;
        }
        else {
            const index = this.indexToKey.length;
            this.indexToKey.push(key);
            this.indexToValue.push(value);
            this.keyToIndex[key] = index;
        }
    }
    delete(key) {
        if (this.has(key)) {
            const lastKey = this.indexToKey.pop();
            const lastValue = this.indexToValue.pop();
            if (key === lastKey) {
                this.keyToIndex[key] = undefined;
            }
            else {
                const index = this.keyToIndex[key];
                this.indexToKey[index] = lastKey;
                this.indexToValue.internalTypedArray[index] = lastValue;
                this.keyToIndex[lastKey] = index;
                this.keyToIndex[key] = undefined;
            }
        }
    }
    clear() {
        this.keyToIndex.length = 0;
        this.indexToKey.length = 0;
        this.indexToValue.clear();
    }
}
