// extra-defer/lib/generator-executor.js
import { goGenerator } from '@blackglory/go';
export class GeneratorExecutor {
    constructor({ autoClear = true } = {}) {
        this.callbacks = [];
        this.autoClear = autoClear;
    }
    get size() {
        return this.callbacks.length;
    }
    defer(callback) {
        this.callbacks.push(callback);
    }
    remove(callback) {
        this.callbacks = this.callbacks.filter(x => x !== callback);
    }
    clear() {
        this.callbacks = [];
    }
    *execute(...args) {
        const callbacks = this.callbacks;
        if (this.autoClear) {
            this.clear();
        }
        for (const callback of this.iterate(callbacks)) {
            yield* goGenerator(() => callback(...args));
        }
    }
    *executeSettled(...args) {
        const callbacks = this.callbacks;
        if (this.autoClear) {
            this.clear();
        }
        for (const callback of this.iterate(callbacks)) {
            try {
                yield* goGenerator(() => callback(...args));
            }
            catch (_a) {
            }
        }
    }
}
