// extra-promise/lib/classes/buffered-channel.js
import { Deferred } from "./deferred.js";
import { DeferredGroup } from "./deferred-group.js";
import { Queue } from '@blackglory/structures';
import { ChannelClosedError } from "../utils/errors.js";
import { FiniteStateMachine } from 'extra-fsm';
export { ChannelClosedError } from "../utils/errors.js";
export class BufferedChannel {
    constructor(bufferSize) {
        this.bufferSize = bufferSize;
        this.fsm = new FiniteStateMachine({
            opened: { close: 'closed' },
            closed: {}
        }, 'opened');
        this.buffer = new Queue();
        this.enqueueDeferredGroup = new DeferredGroup();
        this.dequeueDeferredGroup = new DeferredGroup();
    }
    async send(value) {
        if (this.fsm.matches('closed'))
            throw new ChannelClosedError();
        // 若缓冲区队列已满, 则等待出列信号
        while (this.buffer.size === this.bufferSize) {
            // 双重检查
            if (this.fsm.matches('closed'))
                throw new ChannelClosedError();
            const dequeueDeferred = new Deferred();
            this.dequeueDeferredGroup.add(dequeueDeferred);
            try {
                // 等待出列信号, 如果通道关闭, 则抛出错误
                await dequeueDeferred;
            }
            finally {
                this.dequeueDeferredGroup.remove(dequeueDeferred);
            }
        }
        this.buffer.enqueue(value);
        this.enqueueDeferredGroup.resolve();
    }
    receive() {
        return {
            [Symbol.asyncIterator]: () => {
                if (this.fsm.matches('closed'))
                    throw new ChannelClosedError();
                return {
                    next: async () => {
                        if (this.fsm.matches('closed'))
                            return { done: true, value: undefined };
                        // 缓冲区队列为空, 则等待入列信号
                        while (this.buffer.size === 0) {
                            const enqueueDeferred = new Deferred();
                            this.enqueueDeferredGroup.add(enqueueDeferred);
                            try {
                                // 等待入列信号, 如果通道关闭, 则停止接收
                                await enqueueDeferred;
                            }
                            catch (_a) {
                                return { done: true, value: undefined };
                            }
                            finally {
                                this.enqueueDeferredGroup.remove(enqueueDeferred);
                            }
                        }
                        const value = this.buffer.dequeue();
                        this.dequeueDeferredGroup.resolve();
                        return { done: false, value };
                    },
                    return: async () => {
                        this.close();
                        return { done: true, value: undefined };
                    }
                };
            }
        };
    }
    close() {
        if (this.fsm.matches('opened')) {
            this.fsm.send('close');
            this.buffer.empty();
            this.enqueueDeferredGroup.reject(new ChannelClosedError());
            this.dequeueDeferredGroup.reject(new ChannelClosedError());
        }
    }
}
