// @blackglory/structures/lib/sparse-map.js
export class SparseMap {
    constructor() {
        this.keyToIndex = [];
        this.indexToKey = [];
        this.indexToValue = [];
    }
    get [Symbol.toStringTag]() {
        return this.constructor.name;
    }
    get size() {
        return this.indexToKey.length;
    }
    get internalArray() {
        return this.indexToValue;
    }
    *entries() {
        for (let i = 0; i < this.indexToKey.length; i++) {
            yield [this.indexToKey[i], this.indexToValue[i]];
        }
    }
    keys() {
        return this.indexToKey.values();
    }
    values() {
        return this.indexToValue.values();
    }
    getInternalIndexOfKey(key) {
        return this.keyToIndex[key];
    }
    has(key) {
        return key in this.keyToIndex;
    }
    get(key) {
        if (this.has(key)) {
            const index = this.keyToIndex[key];
            return this.indexToValue[index];
        }
        else {
            return undefined;
        }
    }
    set(key, value) {
        if (this.has(key)) {
            const index = this.keyToIndex[key];
            this.indexToValue[index] = value;
        }
        else {
            const index = this.indexToKey.length;
            this.indexToKey.push(key);
            this.indexToValue.push(value);
            this.keyToIndex[key] = index;
        }
    }
    delete(key) {
        if (this.has(key)) {
            const lastKey = this.indexToKey.pop();
            const lastValue = this.indexToValue.pop();
            if (key === lastKey) {
                delete this.keyToIndex[key];
            }
            else {
                const index = this.keyToIndex[key];
                this.indexToKey[index] = lastKey;
                this.indexToValue[index] = lastValue;
                this.keyToIndex[lastKey] = index;
                delete this.keyToIndex[key];
            }
        }
    }
    clear() {
        this.keyToIndex.length = 0;
        this.indexToKey.length = 0;
        this.indexToValue.length = 0;
    }
}
