// extra-promise/lib/classes/stateful-promise.js
import { FiniteStateMachine } from 'extra-fsm';
export var StatefulPromiseState;
(function (StatefulPromiseState) {
    StatefulPromiseState["Pending"] = "pending";
    StatefulPromiseState["Fulfilled"] = "fulfilled";
    StatefulPromiseState["Rejected"] = "rejected";
})(StatefulPromiseState || (StatefulPromiseState = {}));
export class StatefulPromise extends Promise {
    static from(promise) {
        return new StatefulPromise(async (resolve, reject) => {
            try {
                resolve(await promise);
            }
            catch (e) {
                reject(e);
            }
        });
    }
    get state() {
        return this.fsm.state;
    }
    constructor(executor) {
        const fsm = new FiniteStateMachine({
            [StatefulPromiseState.Pending]: {
                resolve: StatefulPromiseState.Fulfilled,
                reject: StatefulPromiseState.Rejected
            },
            [StatefulPromiseState.Fulfilled]: {},
            [StatefulPromiseState.Rejected]: {}
        }, StatefulPromiseState.Pending);
        super((resolve, reject) => {
            executor(value => {
                fsm.send('resolve');
                resolve(value);
            }, reason => {
                fsm.send('reject');
                reject(reason);
            });
        });
        this.fsm = fsm;
    }
    isPending() {
        return this.fsm.matches(StatefulPromiseState.Pending);
    }
    isFulfilled() {
        return this.fsm.matches(StatefulPromiseState.Fulfilled);
    }
    isRejected() {
        return this.fsm.matches(StatefulPromiseState.Rejected);
    }
}
